File: python/pro14.py
# -*- coding: utf-8 -*-

def rec(m, times):
    if (1 == m):
        return times
    else:
        if (0 == m % 2):
            return rec(m/2, times+1)
        else:
            return rec(3*m+1, times+1)

def collatz_length(n):
    return rec(n, 1)

def pro14(n):
    max = []
    win = 0
    i = 2
    while i <= n:
        len = collatz_length(i)
        if (len > win):
            max = i
            win = len
        i = i + 1
    return max

File: haskell/pro14.hs
collatzLength :: Int -> Int
collatzLength n =
  let rec m times
        | 1 == m = times
        | even m = rec (m `div` 2) (1 + times)
        | otherwise = rec (1 + 3 * m) (1 + times)
  in rec n 1

pro14Helper :: (Int, Int) -> Int -> (Int, Int)
pro14Helper (max, win) i =
  let len = collatzLength i
  in if len > win
     then (i, len)
     else (max, win)

pro14 :: Int -> (Int, Int)
pro14 n =
  foldl pro14Helper (0, 0) [2..n]